### PR TITLE
[TDF][VecOps] Improve integration of VecOps and TDF

### PR DESCRIPTION
**Pull request by bluehood**
Most importantly, now `std::vector<T>` is read as `TVec<T>` in jitted expressions.

### CODE REVIEW

File: tree/treeplayer/src/TDFUtils.cxx [hunk 1]
```diff
@@ -114,23 +114,31 @@ std::string TypeID2TypeName(const std::type_info &id)
...
+   TBranch *branch = tree ? tree->GetBranch(colName.c_str()) : nullptr;
    if (branch) {
       // this must be a real TTree branch
       static const TClassRef tbranchelRef("TBranchElement");
       if (branch->InheritsFrom(tbranchelRef)) {
          // this branch is not a fundamental type, we can ask for the class name
-         return static_cast<TBranchElement *>(branch)->GetClassName();
+         std::string classname(static_cast<TBranchElement *>(branch)->GetClassName());
+         if (classname.compare(0, 7, "vector<") == 0) {
+            // column type is "vector<ValueType>", we read it as "TVec<ValueType>"
```
Comment (dpiparo) [inline tree/treeplayer/src/TDFUtils.cxx]: Maybe here we could leverage the routines in TClassEdit to manipulate names. For example, we could consider:
- TClassEdit::IsSTLCont : in any case, all stl containers are stored the same and are accessible with TTreeReader, therewith TVec 
- TClassEdit::GetSplit: to get out the contained type name

The hard part was to write the code in this PR and surely this is a refinement...

File: tree/treeplayer/src/TDFUtils.cxx [hunk 2]
```diff
@@ -114,23 +114,31 @@ std::string TypeID2TypeName(const std::type_info &id)
...
+            // value type is the classname.size() - 8 chars after the 7th character in "vector<ValueType>"
+            auto valueType = classname.substr(7, classname.size() - 8);
+            colType = "ROOT::Experimental::VecOps::TVec<" + valueType + ">";
```
Comment (dpiparo) [inline tree/treeplayer/src/TDFUtils.cxx]: We build this string over and over: maybe we can add a helper function?
Another detail: we may perhaps avoid copies taking a TVec by reference? In any case the new code allows us to do so!

File: tree/treeplayer/test/CMakeLists.txt [hunk 3]
```diff
@@ -25,8 +25,8 @@ configure_file(dataframe/TCsvDS_test_noheaders.csv . COPYONLY)
 ROOT_ADD_GTEST(datasource_csv dataframe/datasource_csv.cxx LIBRARIES TreePlayer)
 ROOT_ADD_GTEST(datasource_lazy dataframe/datasource_lazy.cxx LIBRARIES TreePlayer)
 
-ROOT_ADD_GTEST(vecops_tvecallocator dataframe/vecops_tadoptallocator.cxx LIBRARIES TreePlayer)
-ROOT_ADD_GTEST(vecops_tvec dataframe/vecops_tvec.cxx LIBRARIES TreePlayer)
+ROOT_ADD_GTEST(vecops_tvecallocator vecops/vecops_tadoptallocator.cxx LIBRARIES TreePlayer)
+ROOT_ADD_GTEST(vecops_tvec vecops/vecops_tvec.cxx LIBRARIES TreePlayer)
```
Comment (dpiparo) [inline tree/treeplayer/test/CMakeLists.txt]: good. In any case these will change location :)

File: tutorials/dataframe/tdf002_dataModel.py [hunk 4]
```diff
@@ -79,22 +79,21 @@
...
-std::vector<double> getPtWeights (const FourVectors &tracks) {
-   std::vector<double> ptsw;
-   ptsw.reserve(tracks.size());
-   for (auto &t : tracks) ptsw.emplace_back(1. / t.Pt());
-   return ptsw;
+using namespace ROOT::Experimental::VecOps;
+TVec<double> getPtWeights(const TVec<FourVector> &tracks)
+{
+   auto ptWeight = [](const FourVector &v) { return 1. / v.Pt(); };
+   return Map(tracks, ptWeight);
```
Comment (dpiparo) [inline tutorials/dataframe/tdf002_dataModel.py]: Excellent: already utilised!

File: tree/treeplayer/src/TDFUtils.cxx (hunk 1 above)
Comment (bluehood) [reply tree/treeplayer/src/TDFUtils.cxx]: sure

File: tree/treeplayer/src/TDFUtils.cxx (hunk 2 above)
Comment (bluehood) [reply tree/treeplayer/src/TDFUtils.cxx]: what do you mean? the string is built ~once per call and it should satisfy all requirements for the c++17 guaranteed NRVO afaik.

The reference after the type name is added when appropriate (always, in the case of TVecs) by other methods, e.g. `JitTransormation`

File: tutorials/dataframe/tdf002_dataModel.py (hunk 4 above)
Comment (bluehood) [reply tutorials/dataframe/tdf002_dataModel.py]: yes, unfortunately this test started crashing after the changes in this PR, and this is not enough to fix it

File: tree/treeplayer/src/TDFUtils.cxx (hunk 2 above)
Comment (bluehood) [reply tree/treeplayer/src/TDFUtils.cxx]: p.s. anyway we should look into avoiding calling `ColumnName2ColumnTypeName` for each column in each transformation/action -- we could build a TDataFrame-wide register that maps columns to types and only perform the lookup once per column per program per TDataFrame graph.

Comment (dpiparo) [reply tree/treeplayer/src/TDFUtils.cxx]: nothing so sophisticated.. Just to centralise the name of the type rather than writing it 3 times, that's all :)

Comment (dpiparo) [reply tree/treeplayer/src/TDFUtils.cxx]: p.s. and invoke it as many times as needed :)

Comment (bluehood) [reply tree/treeplayer/src/TDFUtils.cxx]: ah I understand now! yes, certainly

**Comment by dpiparo** [comment]
I am taking care of this PR, reloading it implementing the changes I myself proposed :-)

**Comment by bluehood** [comment]
Closing as this PR is superseded by #1687

**Pull request closed without merging**